integration/upstream_client_test.go

```diff
@@ -0,0 +1,291 @@
+// +build integration
+
+package integration
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"strconv"
+	"strings"
+	"sync"
+	"testing"
+	"time"
+
+	gcpcachev2 "github.com/envoyproxy/go-control-plane/pkg/cache/v2"
+	gcpserverv2 "github.com/envoyproxy/go-control-plane/pkg/server/v2"
+	gcpserverv3 "github.com/envoyproxy/go-control-plane/pkg/server/v3"
+	gcptest "github.com/envoyproxy/go-control-plane/pkg/test"
+	resourcev2 "github.com/envoyproxy/go-control-plane/pkg/test/resource/v2"
+	gcptestv2 "github.com/envoyproxy/go-control-plane/pkg/test/v2"
+	"github.com/stretchr/testify/assert"
+
+	v2 "github.com/envoyproxy/go-control-plane/envoy/api/v2"
+	corev2 "github.com/envoyproxy/go-control-plane/envoy/api/v2/core"
+	"github.com/envoyproxy/xds-relay/internal/app/upstream"
+	"github.com/envoyproxy/xds-relay/internal/pkg/log"
+)
+
+const (
+	nodeID           = "node-id"
+	originServerPort = 18000
+	loglevel         = "fatal"
+	updates          = 1
+)
+
+var testLogger = log.New(loglevel)
+
+func TestXdsClientGetsIncrementalResponsesFromUpstreamServer(t *testing.T) {
+	updates := 2
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	snapshotsv2, configv2 := createSnapshotCache(updates, testLogger)
+	cb := gcptestv2.Callbacks{Signal: make(chan struct{})}
+	respCh, _, err := setup(ctx, testLogger, snapshotsv2, configv2, &cb)
+	if err != nil {
+		assert.Fail(t, "Setup failed: %s", err.Error())
+		return
+	}
+
+	var wg sync.WaitGroup
+	wg.Add(updates)
+	version := 0
+	go func() {
+		for {
+			select {
+			case r, more := <-respCh:
+				if !more {
+					return
+				}
+				assert.Equal(t, r.VersionInfo, "v"+strconv.Itoa(version))
+				version++
+				wg.Done()
+			}
+		}
+	}()
+
+	sendResponses(ctx, testLogger, updates, snapshotsv2, configv2)
+	wg.Wait()
+
+	timeoutCtx, cancel := context.WithTimeout(ctx, 10*time.Second)
```

Comment:
Since you're reassigning cancel here, does the [first one](https://github.com/envoyproxy/xds-relay/pull/63/files#diff-87e47ce072c022401a8b05fe9def3110R41) get properly cleaned up?

Reply:
hm i didnt notice i was reusing the variable...i will put unique names to avoid confusion.

Reply:
Thanks for catching. done.